src/scheduling/cost_models/simulated_dfs.cc

```diff
@@ -0,0 +1,103 @@
+#include "simulated_dfs.h"
+#include "base/common.h"
+
+#include <algorithm>
+
+#include "misc/SpookyV2.h"
+
+namespace firmament {
+
+// justification for block parameters from Chen, et al (2012)
+// blocks: 64 MB, max blocks 160 corresponds to 10 GB
+SimulatedDFS::SimulatedDFS(uint64_t num_machines, NumBlocks_t blocks_per_machine,
+    uint32_t replication_factor, GoogleBlockDistribution *block_distribution,
+    uint64_t random_seed) :
+    replication_factor(replication_factor), generator(random_seed),
+		uniform(0.0,1.0), blocks_in_file_distn(block_distribution) {
+  NumBlocks_t total_block_capacity = blocks_per_machine * num_machines;
+  total_block_capacity /= replication_factor;
+  LOG(INFO) << "Total capacity of " << total_block_capacity << " blocks.";
+
+  // create files until we hit storage limit
+  while (num_blocks < total_block_capacity) {
+    addFile();
+  }
+  LOG(INFO) << num_blocks << " blocks used, across "
+            << files.size() << " files.";
+}
+
+SimulatedDFS::~SimulatedDFS() { }
+
+void SimulatedDFS::addFile() {
+	uint32_t file_size = numBlocksInFile();  // in blocks
+	num_blocks += file_size;
+	files.push_back(file_size);
+}
+
+uint32_t SimulatedDFS::numBlocksInFile() {
+	double r = uniform(generator);
+	return blocks_in_file_distn->inverse(r);
+}
+
+void SimulatedDFS::addMachine(ResourceID_t machine) {
+  machines.push_back(machine);
+}
+
+void SimulatedDFS::removeMachine(ResourceID_t machine) {
+  for (auto it = machines.begin(); it != machines.end(); ++it) {
```

Comment:
What would the map be keyed by? The machine ID or the resource ID? If we use the machine, there is no benefit over using a vector, since the vector can be subscripted into at O(1) cost. If we use the resource ID, the placement code in GetMachines() will become more complex.

I think I agree that we should change it, but let's quickly discuss the design here.